### PR TITLE
curve-stable-meta-ng: simulator

### PR DESCRIPTION
**Pull request by it4rb**

### CODE REVIEW

File: pkg/liquidity-source/curve/stable-meta-ng/pool_simulator.go [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+	// update balance from the meta swap component
+	metaInfo := swapInfo.Meta
+	t.Reserves[metaInfo.TokenInIndex].Add(&t.Reserves[metaInfo.TokenInIndex], &metaInfo.AmountIn)
+	number.FillBig(&t.Reserves[metaInfo.TokenInIndex], t.Info.Reserves[metaInfo.TokenInIndex])
+
+	t.Reserves[metaInfo.TokenOutIndex].Sub(&t.Reserves[metaInfo.TokenOutIndex], number.Add(&metaInfo.AmountOut, &metaInfo.AdminFee))
+	number.FillBig(&t.Reserves[metaInfo.TokenOutIndex], t.Info.Reserves[metaInfo.TokenOutIndex])
+
+	// if output coin is from base pool
+	withDraw := swapInfo.Withdraw
```
Comment (hieuvo) [inline pkg/liquidity-source/curve/stable-meta-ng/pool_simulator.go:159]: minor typo 
```suggestion
	withdraw := swapInfo.Withdraw
```

File: pkg/liquidity-source/curve/stable-meta-ng/pool_simulator.go [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+	if fromId >= 0 && toId >= 0 {
+		return curve.Meta{
+			TokenInIndex:  fromId,
+			TokenOutIndex: toId,
+			Underlying:    false,
+		}
+	}
+	var baseFromId = t.getUnderlyingIndex(tokenIn)
+	var baseToId = t.getUnderlyingIndex(tokenOut)
+	return curve.Meta{
```
Comment (hieuvo) [inline pkg/liquidity-source/curve/stable-meta-ng/pool_simulator.go:222]: we still return even if 1 of baseFromId and baseToId is negative?

File: pkg/liquidity-source/curve/stable-meta-ng/pool_simulator.go [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+		TokenInIndex:  baseFromId,
+		TokenOutIndex: baseToId,
+		Underlying:    true,
+	}
+}
+
+func (t *PoolSimulator) GetTokens() []string {
+	var result []string
```
Comment (hieuvo) [inline pkg/liquidity-source/curve/stable-meta-ng/pool_simulator.go:230]: minor 
```suggestion
	result := make([]string, 0, len(t.GetInfo().Tokens)+len(t.BasePool.GetInfo().Tokens))
```

File: pkg/liquidity-source/curve/stable-meta-ng/pool_simulator.go (hunk 2 above)
Comment (it4rb) [reply pkg/liquidity-source/curve/stable-meta-ng/pool_simulator.go:222]: this function only get called when encoding route, so it's very likely that the CalcAmountOut has been called and succeeded before with the same parameter, so I think here we can assume that they're correct

yes there will be cases where the route info supplied is not correct, but in those cases the encoded data will likely be incorrect anyway

File: pkg/liquidity-source/curve/stable-meta-ng/pool_simulator.go [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+		return &pool.CalcAmountOutResult{}, ErrTokenToUnderLyingNotSupported
+	}
+
+	if tokenIndexFrom >= 0 && tokenIndexTo >= 0 {
+		// this is normal swap at meta pool, reuse the method from stable-ng
+		return t.PoolSimulator.CalcAmountOut(param)
+	}
+
+	// swap between meta coins and base pool's coins
+	var baseInputIndex = t.basePool.GetTokenIndex(tokenAmountIn.Token)
```
Comment (hieuvo) [inline pkg/liquidity-source/curve/stable-meta-ng/pool_simulator.go:67]: can use just call `t.getUnderlyingIndex`?

File: pkg/liquidity-source/curve/stable-meta-ng/pool_simulator.go [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+func (t *PoolSimulator) CalcAmountOut(param pool.CalcAmountOutParams) (*pool.CalcAmountOutResult, error) {
+	tokenAmountIn := param.TokenAmountIn
+	tokenOut := param.TokenOut
+
+	var tokenIndexFrom = t.Info.GetTokenIndex(tokenAmountIn.Token)
+	var tokenIndexTo = t.Info.GetTokenIndex(tokenOut)
+
+	// cannot swap between the last meta coin and base pool's coins (because the last coin is LPtoken of base pool)
+	if (tokenIndexFrom == t.NumTokens-1 && tokenIndexTo < 0) || (tokenIndexTo == t.NumTokens-1 && tokenIndexFrom < 0) {
+		return &pool.CalcAmountOutResult{}, ErrTokenToUnderLyingNotSupported
```
Comment (hieuvo) [inline pkg/liquidity-source/curve/stable-meta-ng/pool_simulator.go:58]: minor, typo
```suggestion
		return &pool.CalcAmountOutResult{}, ErrTokenToUnderlyingNotSupported
```

File: pkg/liquidity-source/curve/stable-meta-ng/pool_simulator.go (hunk 4 above)
Comment (hieuvo) [inline pkg/liquidity-source/curve/stable-meta-ng/pool_simulator.go:67]: consider to refactor `getUnderlyingIndex` method a little bit, then we can reuse its logic
```golang
func (t *Pool) getUnderlyingIndex(token string) int {
	var tokenIndex = t.GetTokenIndex(token)
	if tokenIndex >= 0 {
		return tokenIndex
	}
	return t.getBaseUnderlyingIndex(tokenIndex, token)
}

func (t *Pool) getBaseUnderlyingIndex(tokenIndex int, token string) int {
	var baseIndex = t.BasePool.GetTokenIndex(token)
	var maxCoin = len(t.Info.Tokens) - 1
	if tokenIndex < 0 && baseIndex >= 0 {
		tokenIndex = maxCoin + baseIndex
	}
	return tokenIndex
}

```

```suggestion
	var baseInputIndex = t.getBaseUnderlyingIndex(tokenIndexFrom, tokenAmountIn.Token)
```

Comment (it4rb) [reply pkg/liquidity-source/curve/stable-meta-ng/pool_simulator.go:67]: um here we need the raw result of `basePool.GetTokenIndex` to determine if both of the coins are from base pool, so I think we can't use `getBaseUnderlyingIndex` here (maybe we can modify it to return a flag to indicate whether it's a base coin, but then it's a bit overcomplicated)

**Pull request merged**